File: WritingState.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Reactive.Subjects;
using RT = clojure.lang.RT;
using Brotherus;

namespace ChineseWriter {

    internal static class WritingState {

        public static Subject<IEnumerable<IDictionary<object, object>>> WordsChanges =
            new Subject<IEnumerable<IDictionary<object, object>>>( );

        // TODO: Don't handle the atom here, make accessors that give parts
        private static IDictionary<object, object> WritingStateData {
            get {
                return (IDictionary<object, object>)
                    ( (clojure.lang.Atom)RT.var( "WritingState", "state" ).deref( ) ).deref( );
            }
        }

        private static IEnumerable<IDictionary<object, object>> Words {
            get {
                return WritingStateData.GetList( "text" ).Cast<IDictionary<object, object>>( );
            }
        }

        /// <summary>
        /// List of pinyin-forms of all words
        /// </summary>
        public static IEnumerable<string> WordPinyins {
            get {
                var words = (IEnumerable<object>) RT.var( "WritingState", "pinyin-text" ).invoke( );
                return words.Cast<string>( );
            }
        }

        public static int CursorPos {
            get {
                return WritingStateData.Get<int>( "cursor-pos" );
            }
        }

        public static void LoadText( ) {
            if ( File.Exists( TextSaveFileName ) ) {
                RT.var( "WritingState", "load-current-text" ).invoke( TextSaveFileName );
            }
            WordsChanges.OnNext( Words );
        }

        public static void Delete( ) {
            DeleteWordInner( CursorPos );
        }

        public static void BackSpace( ) {
            // TODO: Combine these to atomic swap-operation in clojure-side
            DeleteWordInner( CursorPos - 1 );
            Move( "Left" );
        }

        public static void DeleteWordInner( int pos ) {
            RT.var( "WritingState", "delete-word!" ).invoke( pos );
            WordsChanges.OnNext( Words );
        }

        internal static void InsertWord( IHanyuPinyin word ) {
            RT.var( "WritingState", "insert-word!" ).invoke( word.Hanyu, word.Pinyin );
            WordsChanges.OnNext( Words );
        }

        internal static void LiteralInput( string text ) {
            RT.var( "WritingState", "literal-input!" ).invoke( text );
            WordsChanges.OnNext( Words );
        }

        internal static void InsertChinese( string hanyuText ) {
            RT.var( "WritingState", "insert-chinese!" ).invoke( hanyuText );
            WordsChanges.OnNext( Words );
        }

        private static string PadHanyu( string s, int latinLength ) {
            int hanyuExtraLen = s.Replace( " ", "" ).Length / 2;
            return s.PadRight( latinLength - hanyuExtraLen );
        }

        public static string HanyiText( IEnumerable<IDictionary<object, object>> words ) {
            return GetText(words, word => word.Text( ) );
        }

        public static string PinyinText( IEnumerable<IDictionary<object, object>> words ) {
            return GetText( words, word => word.PinyinDiacritics() );
        }

        internal static string GetText( IEnumerable<IDictionary<object, object>> words, Func<IDictionary<object, object>, string> getter ) {
            var finalWords = words != null ? words : Words;
            var texts = finalWords.Select( getter );
            return string.Join( " ", texts.ToArray( ) );
        }

        internal static void Clear( ) {
            RT.var( "WritingState", "clear-current-text!" ).invoke( );
            WordsChanges.OnNext( Words );
        }

        internal static string TextSaveFileName {
            get {
                return Path.Combine( Utils.ExeDir.ToString( ), "text.clj" );
            }
        }

        internal static void SaveCurrentText( ) {
            RT.var( "WritingState", "save-current-text" ).invoke( TextSaveFileName );
        }

        internal static void Move( string dir ) {
            RT.var( "WritingState", "move-cursor!" ).invoke( dir );
            WordsChanges.OnNext( Words );
        }

        internal static void SetCursorPos( int pos ) {
            RT.var( "WritingState", "reset-cursor!" ).invoke( pos );
            WordsChanges.OnNext( Words );
        }

    } // class

} // namespace

File: Clojure/WritingState.clj
(ns WritingState
  (:require [clojure.string :as str])
  (:use utils)
  (:use WordDatabase)
  (:use ParseChinese))

(def state (atom { :text [] :cursor-pos 0 }))

(defn cursor-pos [] (@state :cursor-pos))

(defn current-text [] (@state :text))

(defn clear-current-text! [] (reset! state { :text [] :cursor-pos 0 } ))

(defn pinyin-text [] (map :pinyin (current-text)))

(defn load-current-text [ path ]
  (let [ text (load-from-file path) ]
    (reset! state { :text text :cursor-pos (count text) } )))

(defn save-current-text [ path ]
  (System.IO.File/WriteAllText path (pretty-pr (current-text))))

(defn delete-word [ { :keys [ text cursor-pos ] :as original } delete-pos ]
  (if (and (>= delete-pos 0) (< delete-pos (count text)))
    { :text
     (concat
       (take delete-pos text)
       (drop (inc delete-pos) text))
     :cursor-pos cursor-pos }
    original ))

(defn delete-word! [ pos ]
  (swap! state delete-word pos ))

(defn insert-words [ { :keys [ text cursor-pos ] } new-words ]
  (let [ new-text (concat (take cursor-pos text) new-words (drop cursor-pos text))
         new-cursor-pos (+ cursor-pos (count new-words)) ]
    { :text new-text :cursor-pos new-cursor-pos } ))

(defn insert-words! [ words ] (swap! state insert-words words) )

(defn insert-word!
  ( [ word ] (insert-words! [ word ] ))
  ( [ hanyu pinyin ] (insert-word! { :hanyu hanyu :pinyin pinyin } )))

(defn literal-input! [ text ]
  (insert-word! { :text text } ))

(defn insert-chinese! [ text ]
  (insert-words! (hanyu-to-words text)))

(defn moved-cursor-pos [ { :keys [ text cursor-pos ] } dir ]
  (case dir
    "Left" (max (dec cursor-pos) 0)
    "Right" (min (inc cursor-pos) (count text))
    "Home" 0
    "End" (count text)))

(defn move-cursor [ state dir ]
  (assoc state :cursor-pos (moved-cursor-pos state dir)))

(defn move-cursor! [ dir ]
  (swap! state move-cursor dir ))

(defn set-cursor [ { :keys [ text ] } pos ]
  { :text text :cursor-pos (max 0 (min (count text) pos)) } )

(defn reset-cursor! [ pos ]
  (swap! state set-cursor pos ))
